### PR TITLE
 "add warmup lr_scheduler" create a new pr

### PR DESCRIPTION
**Pull request by solin319**
same as #8005
Sync the code and create a new pr.

### CODE REVIEW

File: python/mxnet/lr_scheduler.py [hunk 1]
```diff
@@ -93,23 +93,27 @@ def __call__(self, num_update):
 
 class MultiFactorScheduler(LRScheduler):
     """Reduce the learning rate by given a list of steps.
-
     Assume there exists *k* such that::
-
-       step[k] <= num_update and num_update < step[k+1]
-
+        step[k] <= num_update and num_update < step[k+1]
```
Comment (szha) [inline python/mxnet/lr_scheduler.py]: In Sphinx, an empty line is needed for code block after `::`

File: python/mxnet/lr_scheduler.py [hunk 2]
```diff
@@ -93,23 +93,27 @@ def __call__(self, num_update):
...
     Then calculate the new learning rate by::
-
-       base_lr * pow(factor, k+1)
-
+        base_lr * pow(factor, k+1)
+    When warmup_step>1, warmup the learning rate by a const value for first warmup_step steps.
+    It returns a new learning rate by::
+        begin_lr + (num_update - 1) * const_update
```
Comment (szha) [inline python/mxnet/lr_scheduler.py]: Same as above.

**Comment by solin319** [comment]
Sorry, I update this branch from apache/master.
Shall I resubmit this pr?
@piiswrong

**Comment by zhreshold** [comment]
@solin319 You can use git rebase, there's no need to close and reopen PRs like that.

File: python/mxnet/lr_scheduler.py [hunk 3]
```diff
@@ -100,16 +100,27 @@ class MultiFactorScheduler(LRScheduler):
...
-        The list of steps to schedule a change
+        The list of steps to schedule a change.
     factor: float
         The factor to change the learning rate.
+    warmup_step : int
+        Changes the learning rate for first 'warmup_step' updates.
+    begin_lr : float, optional
+        The learning rate at begin.
+    stop_lr : float, optional
+        Stop updating the learning rate if it is less than this value.
```
Comment (piiswrong) [inline python/mxnet/lr_scheduler.py]: begin_lr and stop_lr are confusing names and the does doesn't explain exactly what they mean

**Comment by solin319** [comment]
I have  update the parameters name and description.
@piiswrong

**Comment by piiswrong** [comment]
The document still looks cryptic. I don't quite understand why we need 3 more arguments.

@astonzhang @szha Could you help with reviewing the doc?

**Comment by eric-haibin-lin** [comment]
ping @astonzhang @szha

**Comment by szha** [comment]
I think this interface should be redesigned. It's cryptic and hard to understand.

Maybe consider separating the warm-up scheduler as a separate scheduler, and add a class to allow adding two schedulers together. The effect can be that the resulting LR is the sum of the schedulers.

**Comment by solin319** [comment]
@zhreshold @piiswrong @szha 
I combined warmup and multi step scheduler in #8005.
Now, need I separate them?

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!
